src/broker/Data.cc

```diff
@@ -345,6 +345,34 @@ struct val_converter {
 
 			return rval;
 			}
+		else if ( type->Tag() == TYPE_LIST )
+			{
+			// lists are just treated as vectors on the broker side.
+			auto lt = type->AsTypeList();
+			auto rval = make_intrusive<ListVal>(TYPE_ANY);
+			auto types = lt->GetTypes();
+			auto pure = lt->IsPure();
+
+			if ( pure )
+				// ListVal doesn't have a type constructor :)
+				rval = make_intrusive<ListVal>(lt->GetPureType()->Tag());
+
+			unsigned int pos = 0;
+			for ( auto& item : a )
+				{
+				if ( ! pure && pos >= types.size() )
+					return nullptr;
```

Comment:
efficiency nitpick: the bounds check can be done once outside the loop to guard against `! pure && a.size() > types.size()`.